src/liballoc/rc.rs

```diff
@@ -238,22 +241,36 @@ impl<T> Rc<T> {
     /// assert_eq!(Rc::try_unwrap(x), Err(Rc::new(4)));
     /// ```
     #[inline]
-    #[unstable(feature = "rc_unique")]
-    pub fn try_unwrap(rc: Rc<T>) -> Result<T, Rc<T>> {
-        if Rc::is_unique(&rc) {
+    #[unstable(feature = "rc_unique", reason=  "needs FCP", issue = "27718")]
+    pub fn try_unwrap(this: Self) -> Result<T, Self> {
+        if Rc::would_unwrap(&this) {
             unsafe {
-                let val = ptr::read(&*rc); // copy the contained object
-                // destruct the box and skip our Drop
-                // we can ignore the refcounts because we know we're unique
-                deallocate(*rc._ptr as *mut u8, size_of::<RcBox<T>>(),
-                            align_of::<RcBox<T>>());
-                forget(rc);
+                let ptr = *this._ptr;
+                let val = ptr::read(&*this); // copy the contained object
+
+                // remove the implicit "strong weak" pointer now that we've
+                // destroyed the contents.
+                this.dec_weak();
+
+                if this.weak() == 0 {
+                    deallocate(ptr as *mut u8,
+                               size_of_val(&*ptr),
+                               align_of_val(&*ptr))
+                }
```

Comment:
Same comments as above as perhaps this could move to `drop(Weak { ptr: self.ptr })`?